.github/workflows/merge-into-main.yml

```diff
@@ -0,0 +1,48 @@
+name: 'Merge Into Main'
+
+on:
+  push:
+    branches:
+      - main
+
+jobs:
+  terraform-changes:
+    name: Get All Top Level Directories Changed Under `terraform/`
+    outputs:
+      directories: ${{ steps.changed-terraform-directories.outputs.matrix }}
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - run: git fetch origin main:refs/remotes/origin/main
+      - id: changed-terraform-directories
+        run: |
+          top_level_directories_changed=$(git diff HEAD^1 --name-only -- terraform/   `# only compare files changed under the "terraform/" directory` \
+            | sed 's/^terraform\///'                                                       `# removes "terraform/" at the start of the string i.e. "terraform/front-end/main.tf" -> "front-end/main.tf"` \
+            | sed -E 's_/(.*)__'                                                           `# removes everything after the first "/" i.e. "front-end/main.tf" -> "front-end"` \
+            | awk '!seen[$0]++'                                                            `# remove duplicates from list i.e. "front-end\nfront-end" -> "front-end"`\
+            | awk '{ printf("\"%s\",",$0) } END { print s }'                               `# converts multiline string to single line, wrapping each line contents with "", i.e. 'front-end\nbackend' -> '"front-end","backend",'` \
+          )
+          
+          # create a JSON object to be outputted with list of directories
+          echo "matrix={\"matrix\":[${top_level_directories_changed}]}" >> "$GITHUB_OUTPUT"
+
+  terraform-plan:
```

Comment:
copy/paste job here, again will need refactoring in the future.

DIdn't want to add an `apply` step until I can the `HEAD^1` working as expected 👀